src/QsCompiler/Core/Transformations/ExpressionTransformation.fs

```diff
@@ -23,38 +25,40 @@ type ExpressionKindTransformationBase internal (options: TransformationOptions,
 
     let Node = if options.Rebuild then Fold else Walk
 
-    member val internal TypeTransformationHandle = missingTransformation "type" with get, set
     member val internal ExpressionTransformationHandle = missingTransformation "expression" with get, set
 
-    member this.Types = this.TypeTransformationHandle()
-    member this.Expressions = this.ExpressionTransformationHandle()
+    member this.Expressions : ExpressionTransformationBase = this.ExpressionTransformationHandle()
+    member this.Types : TypeTransformationBase = this.ExpressionTransformationHandle().Types
+    member this.Common : CommonTransformationNodes = this.ExpressionTransformationHandle().Types.Common
```

Comment:
So there seems to be a subtle difference in how the various transformations relate to each other as a result of this PR, and I've noticed this change in all the transformation base classes, but we can use the ExpressionKindTransformationBase as an example. The PR changes to minimize the number of different transformation types being passed in. Here for example, we used to have to pass in a type transformation and an expression transformation separately, and these would each be assigned to the expression-kind transformation as fields, allowing the transformation to call into them. But now, both the expression and the type transformation fields are set up using the same transformation passed in: the expression transformation. It uses the expression transformation's Types field instead of an independent type transformation. This is what I'm observing with this change.

Does that limit the flexibility of the transformations? Before, it seems like we would have had the ability to assign one type transformation to the expressions, and a completely separate type transformation to the expression kinds, which might actually make sense in certain circumstances. It seems a shame that we are losing that flexibility with this change.

Reply:
It seems like the major motivation for this change was to allow each of the base transformation classes to have a Common property for referencing the top-level SyntaxTreeTransformation's Common handles, without the need for us to have to pass in a separate Common transformation into each constructor. This means that all transformations in a "bundle" (if I could use that term to describe a SyntaxTreeTransformation and all of its subtransformations) must have the same implementation for, not only their Common handles, but all their handles, removing any modularity the submodules may have had before. One can no longer handle types directly under expression nodes differently from types under expression-kind nodes, for example. Maybe more critically, one can't handle argument tuples differently depending on if you are dealing with a namespace element or an expression. If a developler would need to do make this distinction, they would need to write one handler capable of handling both cases and being able to distinguish between the two based on contextual information that would need to be save and kept track of throughout the transformation "bundle". This can get quite complicated and messy to maintain.

Reply:
Yes, you are right - and great call out! 
Indeed it is a slight change in that now everything has to use the same subtransformation. I believe this is reasonable, since I believe it was rather fragile and error prone before, and the way the different handles where used across transformations wasn't entirely consistent. Also, there seems to be no practical use case anywhere, so I think the simplification is a good change.